lang/src/org/partiql/lang/eval/visitors/AggregateSupportVisitorTransform.kt

```diff
@@ -18,73 +18,46 @@ import com.amazon.ionelement.api.MetaContainer
  * [AggregateRegisterIdMeta].
  *
- * Also collects a list of all aggregate functions used in the `SELECT` list of the current query (excluding subqueries)
+ * Also collects a list of all aggregate functions used in the `SELECT` of the current query (excluding subqueries)
  * and stores them in an instance of [AggregateCallSiteListMeta] on the [PartiqlAst.Expr.Select] instance.
  */
 
-class AggregateSupportVisitorTransform : PartiqlAst.VisitorTransform() {
+class AggregateSupportVisitorTransform : VisitorTransformBase() {
     private val aggregateCallSites = ArrayList<PartiqlAst.Expr.CallAgg>()
 
-    inner class RegisterIdAdderSubVisitorTransform : PartiqlAst.VisitorTransform() {
-        /**
-         * Nests another [AggregateSupportVisitorTransform] within this transform
-         * in order to avoid exposing [aggregateCallSites] to subqueries.
-         */
-        override fun transformExprSelect(node: PartiqlAst.Expr.Select): PartiqlAst.Expr =
-            AggregateSupportVisitorTransform().transformExpr(node)
-
-        override fun transformExprCallAgg(node: PartiqlAst.Expr.CallAgg): PartiqlAst.Expr {
-            aggregateCallSites.add(node)
-
-            return PartiqlAst.build {
-                callAgg_(
-                    setq = node.setq,
-                    funcName = node.funcName,
-                    arg = transformExpr(node.arg),
-                    metas = transformMetas(node.metas) + metaContainerOf(AggregateRegisterIdMeta.TAG to AggregateRegisterIdMeta(aggregateCallSites.size - 1))
-                )
-            }
-        }
-    }
-
-    private val registerIdAdder = RegisterIdAdderSubVisitorTransform()
-
     /**
-     * Applies the [registerIdAdder] only to expressions in select list items.
+     * Nests another [AggregateSupportVisitorTransform] within this transform to avoid operating on
+     * [aggregateCallSites] of a nested `SELECT`.
      */
-    override fun transformProjectItemProjectExpr(node: PartiqlAst.ProjectItem.ProjectExpr): PartiqlAst.ProjectItem =
-        PartiqlAst.build {
-            projectExpr_(
-                expr = registerIdAdder.transformExpr(node.expr),
-                asAlias = node.asAlias,
-                metas = node.metas
-            )
+    override fun transformExprSelect(node: PartiqlAst.Expr.Select): PartiqlAst.Expr {
+        return AggregateSupportVisitorTransform().transformExprSelectEvaluationOrder(node)
+    }
+
+    override fun transformExprCallAgg(node: PartiqlAst.Expr.CallAgg): PartiqlAst.Expr {
+        val transformedCallAgg = PartiqlAst.build {
+            callAgg_(
+                setq = node.setq,
+                funcName = node.funcName,
+                arg = transformExpr(node.arg),
+                metas = transformMetas(node.metas) + metaContainerOf(AggregateRegisterIdMeta.TAG to AggregateRegisterIdMeta(aggregateCallSites.size)))
         }
+        aggregateCallSites.add(transformedCallAgg)
```

Comment:
This rearrangement of the two lines has no impact on the actual logic other than adding the `AggregateRegisterIdMeta` to each `aggregateCallSites`' metas, which will be in the `SELECT` clause's `AggregateCallSiteListMeta` meta.

This made it significantly more clean to test the `AggregateRegisterIdMeta` for the right value in testing.